### DISCUSSION TITLE
Feature Request: Ability to Set Tab Color Programmatically

### DISCUSSION THREAD

**Discussion by Niwotian60** [microsoft/terminal#16490] [Ideas]
This might be a very niche request, but I have a use case that would really benefit from having the ability to set the tab color from a shell script. I often run builds or tests that take a fair bit of time to complete and being able to set the tab color of a background tab dynamically based on the outcome would be very useful. 

Currently I set the title dynamically using  `$Host.UI.RawUI.WindowTitle = $title`  but it lacks the impact of turning a tab bright green or red when something succeeds or fails.  There is the ability to set individual tab colors when launching a new tab, so it seems like a strange omission that there isn't a similar dynamic `$Host.UI.RawUI.WindowColor = $color` capability.

**Comment by 237dmitry** [comment]
If shell can work with json then this is easy to do:

![Screenshot 2023-12-21 201830](https://github.com/microsoft/terminal/assets/78153320/f250ae5b-d03a-4ae6-91b7-0e4b0713b344)

**Comment by Niwotian60** [comment]
Dmitry, Thank you for the quick response.

I'm not by any means a PowerShell expert (I use it out of necessity) but it does have json support.  Before I hack something up, I want to be sure I understand this. Wouldn't changing the profile change EVERY powershell tab color? I typically have six instances open in Windows Terminal (two split windows in three separate tabs). I don't want to update the tabs of every instance of that profile, and while I could create a separate profile for each tab, that comes with its own management issues.

**Comment by 237dmitry** [comment]
> Wouldn't changing the profile change EVERY powershell tab color?

No, it will change tabColor of all of profile instances.

**Reply by 237dmitry** [reply]
But you can start new profile instance with `--tabColor` parameter:

![Screenshot 2023-12-21 205304](https://github.com/microsoft/terminal/assets/78153320/f7071023-199a-4455-947a-bea7c42df27d)

```powershell
$colors = "#ff0000","#00ff00","#0000ff"
0..2 | % { wt -w 0 nt --tabColor $colors[$_] -p "PowerShell" }
```

**Comment by Niwotian60** [comment]
I already do use the launch setting. That's why I thought it strange there was no way to change the color dynamically. This is my current terminal; note the model run on the second tab. I have my tabs all set to launch with colors appropriate to their content (purple for twitch, blue for python, rust for rust, orange for ubuntu, etc.) Because colors are good!
![win_term](https://github.com/microsoft/terminal/assets/5266833/8678bb51-2925-4025-8ddf-4cd2266cdded)

**Comment by j4james** [comment]
You can use an escape sequence to set the tab color of the active tab. For example, in PowerShell, you can make it red with:
```
"`e[2;15;1,|"
```
In bash, it would be something like this:
```
printf "\e[2;15;1,|"
```

It's the last number in that sequence that specifies the background color: 1 is red, 2 is green, 3 is yellow, etc. The 15 represents the foreground color, but doesn't currently have any effect in Windows Terminal (the foreground is automatically assigned a contrasting color).

**Reply by 237dmitry** [reply]
> "`e[2;15;1,|"

Is there a sequence for 256 and rgb palettes?

**Reply by j4james** [reply]
See https://github.com/MicrosoftDocs/Console-Docs/issues/267

**Reply by pelamfi** [reply]
This discussion was one of the first google hits. I ended up writing this SU post with more details about this. I hope you don't mind me crosslinking it here to help fellow wt.exe users wondering about this stuff:

https://superuser.com/questions/1909260/how-to-change-windows-terminal-wt-exe-tab-colors-and-default-foreground-and-ba/1909261#1909261

**Comment by Niwotian60** [comment]
[Edit] I figured out the problem. Apparently if the tab is launched with the --tabColor explicitly set it can't be overridden with escape commands. If I simply launch a new shell without them it works fine, although that hampers my standard launch which is a powershell script that takes the type of shell I want to launch and colorizes it, changes directory, sets the posh line, etc. The main launch sequence is:
```
wt.exe ` -p $s1.profile --title "$($s1.title)" --tabColor $s1.tab_color -d $s1.start_dir $options1 `; `
         split-pane `
         -p $s2.profile --title "$($s2.title)" --tabColor $s2.tab_color -d $s2.start_dir $options2 `; `
         move-focus left
```
Curses? Wow, that takes me ... way ... back!

@j4james I tried your curses commands in both PowerShell and linux with no success. I tried a few other curses commands such as "`e[34;42m" which correctly changed my fg/bg to black on green so it doesn't appear to be a configuration issue. I'm running Windows Terminal Preview Version: 1.19.3172.0 if that matters.

**Reply by j4james** [reply]
> Apparently if the tab is launched with the --tabColor explicitly set it can't be overridden with escape commands.

@Niwotian60 Feel free to open an issue for that, if you think the escape sequences should be able to override the command line parameters. That doesn't seem like an unreasonable request, given that you can use escape sequences to override a tab title that has been set via the command line. I'm actually somewhat to blame for the current order of precedence, but in retrospect I think it's wrong.

**Reply by Niwotian60** [reply]
@j4james First, I don't believe anyone is to "blame". You are providing your time and skills to support a very nice terminal emulator and made a design decision. But rethinking our decisions and refactoring our code is part of the process.

If the best path forward is to open an issue I can certainly do that, but my original feature request is still my preference. Having a simple API such as: `setTitleColor(fg:Color, bg:Color) -> (oldFgColor, oldBgColor)` is far more easily understood and useable than a 50-year old terminal control sequence. And with such an API, adding support for such a sequence would likely be a simple protocol mapping.

Is there a path to open an "issue" for a feature request? I would even be willing to help implement such a feature depending on the underlying codebase (noting that I'm not a Windows developer and don't have knowledge of .NET).

**Reply by j4james** [reply]
@Niwotian60 If I'm understanding you correctly, what you're asking for is a shell API (like `$Host.UI.RawUI.WindowTitle`), but that's not something provided by the terminal. You'd probably need to ask in the [PowerShell](https://github.com/PowerShell/PowerShell) repository. Ultimately they'd just be converting your API call into an escape sequence, though, so it would be subject to the same limitations (i.e. the command line parameter taking precedence).
